// value-store/lib/index.js
var ValueStore, corepath, fs, ini, readIni, readJson, setSource;

fs = require('fs');

corepath = require('path');

ini = require('ini');

readJson = function(file) {
  return JSON.parse(fs.readFileSync(file, 'utf8'));
};

readIni = function(file) {
  return ini.parse(fs.readFileSync(file, 'utf8'));
};

setSource = function(object, value) {
  Object.defineProperty(object, '__source', {
    value: value,
    writable: false,
    enumerable: false,
    configurable: false
  });
};

ValueStore = (function() {
  function ValueStore(options, validated) {
    var array, element, i, index, len, ref;
    if (!validated) {
      array = Array.isArray(options) ? options : (ref = options.array) != null ? ref : [];
      for (index = i = 0, len = array.length; i < len; index = ++i) {
        element = array[index];
        if (typeof element !== 'object') {
          throw new Error('ValueStore accepts only objects. Invalid element at ' + index + ': ' + element);
        } else {
          if (element.__source == null) {
            setSource(element, 'constructor');
          }
        }
      }
    } else {
      array = options.array;
    }
    this.array = array;
  }

  ValueStore.prototype.count = function() {
    return this.array.length;
  };

  ValueStore.prototype.source = function(index) {
    var ref, ref1;
    if (index == null) {
      index = 0;
    }
    return (ref = this.array) != null ? (ref1 = ref[index]) != null ? ref1.__source : void 0 : void 0;
  };

  ValueStore.prototype.get = function(key, index) {
    var ref, ref1;
    if (index != null) {
      return (ref = this.array) != null ? (ref1 = ref[index]) != null ? ref1[key] : void 0 : void 0;
    }
    index = this["in"](key);
    if (index > -1) {
      return this.array[index][key];
    } else {
      return void 0;
    }
  };

  ValueStore.prototype.has = function(key, index) {
    var ref, ref1;
    if (index != null) {
      return ((ref = this.array) != null ? (ref1 = ref[index]) != null ? ref1[key] : void 0 : void 0) != null;
    } else {
      return this["in"](key) > -1;
    }
  };

  ValueStore.prototype["in"] = function(key) {
    var i, index, len, object, ref;
    ref = this.array;
    for (index = i = 0, len = ref.length; i < len; index = ++i) {
      object = ref[index];
      if (object[key] != null) {
        return index;
      }
    }
    return -1;
  };

  ValueStore.prototype.info = function(key, index) {
    var ref, ref1, result;
    if (index == null) {
      index = this["in"](key);
    }
    return result = {
      "in": index,
      key: key,
      value: (ref = this.array) != null ? (ref1 = ref[index]) != null ? ref1[key] : void 0 : void 0,
      overridden: this._isOverridden(key, index)
    };
  };

  ValueStore.prototype._isOverridden = function(key, index) {
    var i, ref, ref1, ref2;
    if (index < 1) {
      return false;
    }
    if (index > this.array.length) {
      index = this.array.length;
    }
    for (index = i = 0, ref = index; 0 <= ref ? i <= ref : i >= ref; index = 0 <= ref ? ++i : --i) {
      if (((ref1 = this.array) != null ? (ref2 = ref1[index]) != null ? ref2[key] : void 0 : void 0) != null) {
        return true;
      }
    }
    return false;
  };

  ValueStore.prototype.all = function(key) {
    var i, len, object, ref, result;
    result = [];
    ref = this.array;
    for (i = 0, len = ref.length; i < len; i++) {
      object = ref[i];
      if (object[key] != null) {
        result.push(object[key]);
      }
    }
    return result;
  };

  ValueStore.prototype.add = function(key, value, index) {
    if (index == null) {
      index = 0;
    }
    return this.set(key, value, index, true);
  };

  ValueStore.prototype.remove = function(key, index) {
    return this.set(key, void 0, index, false);
  };

  ValueStore.prototype.set = function(key, value, index, add) {
    var existingValue, ref, ref1, result;
    if (!(this.array.length > 0)) {
      return {
        error: 'Invalid index: ' + (index != null ? index : 0)
      };
    }
    if (key == null) {
      return {
        error: 'No key specified'
      };
    }
    if (!(add === false || (value != null))) {
      return {
        error: 'No value specified'
      };
    }
    if (index != null) {
      if (!(index > -1 && index < this.array.length)) {
        return {
          error: 'Invalid index: ' + index
        };
      }
    } else if (add === false) {
      index = this["in"](key);
      if (index === -1) {
        return;
      }
    } else {
      index = 0;
    }
    existingValue = (ref = this.array) != null ? (ref1 = ref[index]) != null ? ref1[key] : void 0 : void 0;
    result = {};
    if (add === true) {
      if (existingValue != null) {
        if (Array.isArray(existingValue)) {
          existingValue.push(value);
        } else {
          this.array[index][key] = [existingValue, value];
        }
      } else {
        this.array[index][key] = value;
      }
      result.addedTo = existingValue;
    } else if (add === false) {
      result.removed = existingValue;
      this.array[index][key] = void 0;
    } else {
      result.replaced = existingValue;
      this.array[index][key] = value;
    }
    return result;
  };

  ValueStore.prototype.append = function(thing, options) {
    return this._insert(false, thing, options);
  };

  ValueStore.prototype.prepend = function(thing, options) {
    return this._insert(true, thing, options);
  };

  ValueStore.prototype._insert = function(first, thing, options) {
    var error, object, parse;
    switch (typeof thing) {
      case 'string':
        thing = corepath.resolve(thing);
        parse = thing.slice(-5) === '.json' || (options != null ? options.format : void 0) === 'json' ? readJson : thing.slice(-4) === '.ini' || (options != null ? options.format : void 0) === 'ini' ? readIni : null;
        if (parse == null) {
          return {
            error: 'String must be a json or ini file'
          };
        }
        try {
          object = parse(thing);
          if (object.__source == null) {
            setSource(object, {
              file: thing,
              format: parse === readJson ? 'json' : 'ini',
              fn: first ? 'prepend' : 'append'
            });
          }
        } catch (error1) {
          error = error1;
          if (error.code === 'MODULE_NOT_FOUND' || error.code === 'ENOENT') {
            return {
              error: 'File doesn\'t exist: ' + thing,
              exists: false
            };
          }
          return {
            error: 'Failed to require file',
            reason: error
          };
        }
        break;
      case 'object':
        object = thing;
        if (object.__source == null) {
          setSource(object, (first ? 'prepend' : 'append'));
        }
        break;
      default:
        return {
          error: 'Must provide a string or object',
          value: thing
        };
    }
    if (first) {
      this.array.unshift(object);
    } else {
      this.array.push(object);
    }
    return true;
  };

  ValueStore.prototype.shift = function(count) {
    var result;
    if (count == null) {
      count = 1;
    }
    if (count < 1) {
      return {
        removed: []
      };
    }
    result = {
      removed: this.array.slice(0, count)
    };
    if (count > 1) {
      this.array.splice(0, count);
    } else {
      this.array.shift();
    }
    return result;
  };

  ValueStore.prototype.pop = function(count) {
    var result;
    if (count == null) {
      count = 1;
    }
    if (count < 1) {
      return {
        removed: []
      };
    }
    result = {
      removed: this.array.slice(-count)
    };
    if (count > 1) {
      this.array.splice(-count, count);
    } else {
      this.array.pop();
    }
    return result;
  };

  ValueStore.prototype.write = function(index, options) {
    var error, ext, file, object, ref, source, stringify;
    if (index == null) {
      index = 0;
    }
    if (!(index > -1 && index < this.array.length)) {
      return {
        error: 'Invalid index: ' + index
      };
    }
    object = this.array[index];
    source = this.source(index);
    file = (ref = options != null ? options.file : void 0) != null ? ref : source != null ? source.file : void 0;
    if (file == null) {
      return {
        error: 'No `file` in source #' + index + ' or in options'
      };
    }
    ext = corepath.extname(file);
    stringify = (options != null ? options.format : void 0) === 'ini' || source.format === 'ini' || ext === '.ini' ? ini.stringify.bind(ini, object, {
      whitespace: true
    }) : JSON.stringify.bind(JSON, object, null, 2);
    try {
      fs.writeFileSync(file, stringify() + '\n', 'utf8');
    } catch (error1) {
      error = error1;
      return {
        error: 'Failed to write source #' + index,
        reason: error
      };
    } finally {
      setSource(object, source);
    }
  };

  return ValueStore;

})();

module.exports = function(options) {
  var array, element, i, index, len;
  if (options != null) {
    if (Array.isArray(options)) {
      array = options;
      options = {
        array: array
      };
    } else if (options.array != null) {
      array = options.array;
    } else {
      array = [];
      options.array = array;
    }
  } else {
    array = [];
    options = {
      array: array
    };
  }
  for (index = i = 0, len = array.length; i < len; index = ++i) {
    element = array[index];
    if (typeof element !== 'object') {
      return {
        error: 'ValueStore accepts only objects. Invalid element at ' + index + ': ' + element
      };
    } else {
      setSource(element, 'constructor');
    }
  }
  return new ValueStore(options, true);
};

module.exports.ValueStore = ValueStore;
