// nuc/bin/do-set.js

/*
This is the 'set' subcommand file for the `nuc` command line tool.

It's responsible for:
  1. using the `nuc` lib to change the desired configuration
     a. set - will insert or update the key/value
     b. add - do-add calls this with option add:true, values are combined
     c. delete - do-delete call this with option delete:true, key is removed
  2. pretty printing the change results to the console

Format:  nuc [@id] set key value [<scope>]

The '@id' is only required if there isn't a .nuc.name file to get the id from.

Examples:

  These all exclude the ID to make them shorter, but, they could have @id in there.

  nuc set somekey somevalue - create/update key in first writable config file found
                              or, create one in the default scope, 'user'
  nuc set somekey somevalue local  - same as above, but only in the 'local' scope
  nuc set somekey somevalue user   - same as above, but only in the 'local' scope
  nuc set somekey somevalue system - same as above, but only in the 'local' scope

Writable scopes:

1. local - possibly, if the user has the right permissions
2. user - the user has permissions for its own home directory
3. system - depends on the user's permissions
 */
module.exports = function(options) {
  var command, key, ref, ref1, result, scope, value;
  key = options.args[0];
  if (!options.remove) {
    value = options.args[1];
  }
  command = (ref = options.name) != null ? ref : 'set';
  if (!(options.remove || (value != null))) {
    if (key == null) {
      return {
        error: "Must specify <key> and <value> for " + command + " command"
      };
    } else {
      return {
        error: "Must specify <value> for " + command + " command"
      };
    }
  }
  if (key == null) {
    return {
      error: 'Must specify <key> for set command'
    };
  }
  scope = options.remove ? options.args[1] : options.args[2];
  result = require('../lib').set({
    id: options.id,
    key: key,
    value: value,
    scope: scope,
    add: options.add,
    remove: options.remove,
    defaults: options.defaults
  });
  if ((result != null ? result.error : void 0) != null) {
    return result;
  }
  console.log("app @" + options.id + " " + (scope != null ? 'in scope ' + scope : ''));
  if (options.remove) {
    if (result.replaced != null) {
      console.log("  removed \'" + key + "\'");
    } else {
      console.log("Unable to remove '" + key + "' because it didn\'t exist");
    }
  } else {
    console.log("  '" + key + "'", '=', (ref1 = result.newValue) != null ? ref1 : value);
  }
  if ((result != null ? result.replaced : void 0) != null) {
    console.log("  '" + key + "'", 'was', result.replaced);
  }
};
