// nuc/lib/get.js
var findId,
  indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

findId = require('./find-id');

module.exports = function(options) {
  var buildValueStore, deepExtend, gather, gatherers, i, input, j, len, len1, object, ref, ref1, ref2, result, store, values;
  if (options == null) {
    options = {};
  }
  deepExtend = require('deep-extend');
  buildValueStore = require('value-store');
  gatherers = [require('./from-files'), require('./from-env'), require('./from-cli')];
  result = findId(options);
  if ((result != null ? result.__error : void 0) != null) {
    return result;
  }
  store = buildValueStore({
    array: options.defaults != null ? [options.defaults] : []
  });
  if (options.collapse !== false) {
    values = (ref = options.defaults) != null ? ref : {};
  }
  input = {
    id: options.id,
    scope: options.scope,
    values: values,
    store: store,
    each: function(file, exists) {
      if (exists) {
        result = store.prepend(file);
        if (values != null) {
          deepExtend(values, store.array[0]);
        }
        if (!options.stack) {
          store.array.shift();
        }
      }
    }
  };
  for (i = 0, len = gatherers.length; i < len; i++) {
    gather = gatherers[i];
    if ((typeof scope !== "undefined" && scope !== null) && (ref1 = !scope, indexOf.call(gather.scopes, ref1) >= 0)) {
      continue;
    }
    result = gather(input);
    if (result.error != null) {
      return result;
    }
    ref2 = result.objects;
    for (j = 0, len1 = ref2.length; j < len1; j++) {
      object = ref2[j];
      if (!(Object.keys(object).length > 0)) {
        continue;
      }
      if (values != null) {
        deepExtend(values, object);
      }
      if (options.stack) {
        store.prepend(object);
      }
    }
  }
  if (values != null) {
    delete values.__source;
  }
  return result = values != null ? options.stack != null ? {
    values: values,
    store: store
  } : values : store;
};
