// nuc/lib/set.js
var findId;

findId = require('./find-id');

module.exports = function(options) {
  var closestExistingFile, corepath, createScope, getFiles, home, newValue, op, ref, ref1, ref2, ref3, ref4, ref5, replaced, result, searchScope, store, writeOptions;
  result = findId(options);
  if ((result != null ? result.__error : void 0) != null) {
    return result;
  }
  corepath = require('path');
  getFiles = require('./get-files');
  if (options.scope != null) {
    if ((ref = options.scope) !== 'local' && ref !== 'user' && ref !== 'system') {
      return {
        error: 'writable scopes: local, user, system. not: ' + options.scope
      };
    }
    searchScope = createScope = options.scope;
  } else {
    searchScope = ['local', 'user', 'system'];
    createScope = (ref1 = options.createScope) != null ? ref1 : options.defaults.createFile.scope;
  }
  store = require('value-store')();
  result = getFiles({
    id: options.id,
    scope: searchScope
  });
  if ((result != null ? (ref2 = result.files) != null ? ref2.length : void 0 : void 0) > 0) {
    closestExistingFile = result.files[result.files.length - 1];
  }
  store.prepend(closestExistingFile != null ? closestExistingFile : {});
  op = options.add ? 'add' : options.remove ? 'remove' : 'set';
  if (op !== 'add') {
    replaced = store.get(options.key);
  }
  store[op](options.key, options.value);
  if (op !== 'remove') {
    newValue = store.get(options.key);
  }
  if (closestExistingFile == null) {
    writeOptions = {
      format: (ref3 = options.createFormat) != null ? ref3 : options.defaults.createFile.format,
      file: (ref4 = options.createFile) != null ? ref4 : options.defaults.createFile[createScope]
    };
    if (writeOptions.file[0] === '~') {
      home = (ref5 = process.env.HOME) != null ? ref5 : process.env.USERPROFILE;
      writeOptions.file = corepath.join(home, writeOptions.file.slice(1));
    }
    writeOptions.file = writeOptions.file.replace('{id}', options.id);
  }
  result = store.write(0, writeOptions);
  return {
    replaced: replaced,
    newValue: newValue
  };
};
